pkg/core/common/auth/token.go

```diff
@@ -11,9 +13,31 @@ type TransportWithToken struct {
 	Token string
 }
 
+func RefreshServiceAuth() (string, error) {
```

Comment:
There are issues with this implementation given https://pkg.go.dev/net/http#RoundTripper:

```
A RoundTripper must be safe for concurrent use by multiple goroutines.
```

```
	// RoundTrip should not attempt to interpret the response. In
	// particular, RoundTrip must return err == nil if it obtained
	// a response, regardless of the response's HTTP status code.
	// A non-nil err should be reserved for failure to obtain a
	// response. Similarly, RoundTrip should not attempt to
	// handle higher-level protocol details such as redirects,
	// authentication, or cookies.
```

For k8s, it's not guaranteed that these clients will be using service account auth so I think the logic should be moved to the actual client.

This token helper is also used for the prometheus exporter client so would be problematic in this context.

Reply:
1. Off the cuff I'd assume `DoRequestAndSetValue()` is the place to update the transport's token (w/ RWMutex?).
2. The [file config source](https://github.com/signalfx/signalfx-agent/blob/main/pkg/core/config/sources/file/file.go) has logic for watching for changes but its interface is pretty tied to config content so I don't think would be directly usable as is. The diff detection is by comparing checksums and we could set up a similar loop goroutine in client creation.